File: task-based-experiment/benchmarks/600.opentelemetry/630.thumbnailer-opentelemetry/nodejs/function.js
const sharp = require('sharp');
const path = require('path');
const storage = require('./storage');

const { BasicTracerProvider } = require('@opentelemetry/sdk-trace-base');
const { Resource } = require('@opentelemetry/resources');
const { SEMRESATTRS_SERVICE_NAME } = require('@opentelemetry/semantic-conventions');
const { OTLPTraceExporter } = require('@opentelemetry/exporter-trace-otlp-grpc');
const { SimpleSpanProcessor } = require('@opentelemetry/sdk-trace-base');
const opentelemetry = require('@opentelemetry/api');

const tracerProvider = new BasicTracerProvider({
  resource: new Resource({
    [SEMRESATTRS_SERVICE_NAME]: '630.thumbnailer-opentelemetry-nodejs',
  }),
});

const traceExporter = new OTLPTraceExporter({
  url: 'http://192.168.1.104:4317',
});

tracerProvider.addSpanProcessor(new SimpleSpanProcessor(traceExporter));
tracerProvider.register();

const tracer = tracerProvider.getTracer('nodejs-tracer');

let storage_handler = new storage.storage();

exports.handler = async function(event) {
  const span = tracer.startSpan('handler');

  const bucket = event.bucket.bucket;
  const input_prefix = event.bucket.input;
  const output_prefix = event.bucket.output;
  const key = event.object.key;
  const width = event.object.width;
  const height = event.object.height;

  let pos = key.lastIndexOf('.');
  let upload_key = key.substr(0, pos < 0 ? key.length : pos) + '.png';

  span.setAttributes({
    'bucket': bucket,
    'input_prefix': input_prefix,
    'output_prefix': output_prefix,
    'key': key,
    'width': width,
    'height': height,
    'upload_key': upload_key,
  });

  const sharp_resizer = sharp().resize(width, height).png();
  let read_promise = storage_handler.downloadStream(bucket, path.join(input_prefix, key));
  span.addEvent('Download stream created');

  let [writeStream, promise, uploadName] = storage_handler.uploadStream(bucket, path.join(output_prefix, upload_key));
  span.setAttribute("upload.uploadName", uploadName);
  span.addEvent('Upload stream created');

  read_promise.then(
    (input_stream) => {
      span.addEvent('Input stream received');
      input_stream.pipe(sharp_resizer).pipe(writeStream);
      span.addEvent('Piping streams');
    }
  ).catch((err) => {
    span.recordException(err);
    span.end();
  });

  await promise.then(() => {
    span.addEvent('Upload completed');
  }).catch((err) => {
    span.recordException(err);
  });

  span.end();
  return { bucket: output_prefix, key: uploadName };
};

File: e2-task-based-experiment/benchmarks/600.opentelemetry/6300.thumbnailer-opentelemetry/python/function.py
import datetime
import io
import os
import sys
import uuid
from urllib.parse import unquote_plus
from PIL import Image

from . import storage
client = storage.storage.get_instance()

from opentelemetry import trace
from opentelemetry.exporter.otlp.proto.grpc.trace_exporter import OTLPSpanExporter
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export import SimpleSpanProcessor
from opentelemetry.sdk.resources import Resource

resource = Resource(attributes={
    "service.name": "630.thumbnailer-opentelemetry"
})

trace.set_tracer_provider(TracerProvider(resource=resource))
tracer = trace.get_tracer("function")

otlp_exporter = OTLPSpanExporter(
    endpoint="http://<REPLACE_ME>:4317",
    insecure=True
)

span_processor = SimpleSpanProcessor(otlp_exporter)
trace.get_tracer_provider().add_span_processor(span_processor)

def resize_image(image_bytes, w, h, ctx):
    with tracer.start_as_current_span("resize_image", ctx) as span:
        span.set_attribute("width", w)
        span.set_attribute("height", h)
        
        with Image.open(io.BytesIO(image_bytes)) as image:
            image.thumbnail((w, h))
            out = io.BytesIO()
            image.save(out, format='jpeg')
            out.seek(0)
            resized_size = out.getbuffer().nbytes
            span.set_attribute("resized_size", resized_size)
            return out

def handler(event):
    span = tracer.start_span("handler")
    ctx = trace.set_span_in_context(span)

    bucket = event.get('bucket').get('bucket')
    input_prefix = event.get('bucket').get('input')
    output_prefix = event.get('bucket').get('output')
    key = unquote_plus(event.get('object').get('key'))
    width = event.get('object').get('width')
    height = event.get('object').get('height')

    span.set_attribute("bucket", bucket)
    span.set_attribute("input_prefix", input_prefix)
    span.set_attribute("output_prefix", output_prefix)
    span.set_attribute("key", key)
    span.set_attribute("width", width)
    span.set_attribute("height", height)

    download_span = tracer.start_span("download", context=ctx)
    img = client.download_stream(bucket, os.path.join(input_prefix, key))
    download_span.end()

    resize_span = tracer.start_span("resize", context=ctx)
    resized = resize_image(img, width, height, ctx)
    resized_size = resized.getbuffer().nbytes
    resize_span.end()

    upload_span = tracer.start_span("upload", context=ctx)
    key_name = client.upload_stream(bucket, os.path.join(output_prefix, key), resized)
    upload_span.end()

    span.set_attribute("download_size", len(img))
    span.set_attribute("upload_size", resized_size)
    span.end()

    return {
        'result': {
            'bucket': bucket,
            'key': key_name
        }
    }
